[Cocoa] Support AVIF images for macOS Ventura and iOS 16

shallawa opened the pull request:
#### 59d774a32243ae8ba59de9178f3b029fd28c4477
<pre>
[Cocoa] Support AVIF images for macOS Ventura and iOS 16
<a href="https://bugs.webkit.org/show_bug.cgi?id=241904">https://bugs.webkit.org/show_bug.cgi?id=241904</a>
rdar://95742091

Reviewed by Myles C. Maxfield.

Add the mime type and the UTI of the AVIF to the list of the allowed image
formats. The system frameworks will be used to render the AVIF images on macOS
Ventura and iOS 16. Because of sand-boxing limitations, software decoding has
to be used for AVIF images.

* LayoutTests/platform/mac/TestExpectations:
* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/loader/cache/CachedResourceRequest.cpp:
(WebCore::acceptHeaderValueForImageResource):
* Source/WebCore/platform/MIMETypeRegistry.cpp:
* Source/WebCore/platform/graphics/cg/ImageDecoderCG.cpp:
(WebCore::createImageSourceOptions):
* Source/WebCore/platform/graphics/cg/UTIRegistry.cpp:
(WebCore::defaultSupportedImageTypes):

Canonical link: <a href="https://commits.webkit.org/251850@main">https://commits.webkit.org/251850@main</a>
</pre>

litherum commented on Source/WebCore/loader/cache/CachedResourceRequest.cpp:
```diff
@@ -126,17 +126,23 @@ void CachedResourceRequest::setDomainForCachePartition(const String& domain)
 #else
     #define WEBP_HEADER_PART ""
+#endif
+#if HAVE(AVIF) || USE(AVIF)
```
This is kind of turning into soup. Maybe if we pulled this out into helper functions it would be more readable?

shallawa replied:
The strings for the WebP and AVIF mime types are moved to separate functions.

litherum commented on Source/WebCore/platform/MIMETypeRegistry.cpp:
```diff
@@ -93,7 +93,7 @@ constexpr ComparableCaseFoldingASCIILiteral supportedImageMIMETypeArray[] = {
     "image/apng",
 #endif
-#if USE(AVIF)
+#if HAVE(AVIF) || USE(AVIF)
```
uh oh. Why is there both `HAVE(AVIF)` and `USE(AVIF)`? Can we consolidate them?

shallawa replied:
HAVE(AVIF) is used by Apple ports where the system frameworks provide the needed support to render these images.
USE(AVIF) is used by non Apple ports. It is defined in Source/cmake/OptionsWPE.cmake and Source/cmake/OptionsGTK.cmake. They use libavif to handle the AVIF images. If we want to merge HAVE(AVIF) and USE(AVIF) and merge HAVE(WEBP) and USE(WEBP), I think it is better to handle this in a separate patch.

litherum replied:
Can we make sure we have a bug open for the separate patch? (Presumably the separate patch would be for both AVIF and WEBP.)

shallawa replied:
https://bugs.webkit.org/show_bug.cgi?id=241997

litherum commented on Source/WebCore/platform/MIMETypeRegistry.cpp:
```diff
@@ -133,7 +133,7 @@ constexpr ComparableCaseFoldingASCIILiteral supportedImageMIMETypeArray[] = {
     "image/vnd.switfview-jpeg",
 #endif
-#if (USE(CG) && HAVE(WEBP)) || (!USE(CG) && USE(WEBP))
+#if HAVE(WEBP) || USE(WEBP)
```
🤔

shallawa replied:
The definition of HAVE_WEBP in PlatfromHave.h states that it is enabled on Apple ports only

```
#if PLATFORM(COCOA)
#define HAVE_WEBP 1
#endif
```

The definition of USE_WEBP in PlatfromUse.h states that it is enabled on GTK and WPE.

```
#if PLATFORM(GTK) || PLATFORM(WPE)
#define USE_WEBP 1
#endif
```

It is also defined in Source/cmake/OptionsPlayStation.cmake and Source/cmake/OptionsWinCairo.cmake.

So I think there is no need to check for the platform also.

litherum commented on Source/WTF/wtf/PlatformHave.h:
```diff
@@ -1265,6 +1265,14 @@
+
 #if !PLATFORM(IOS_FAMILY)
 #define HAVE_MEDIA_VOLUME_PER_ELEMENT 1
 #endif
```
(not sure where to leave this comment) It's pretty surprising that no sandbox changes are necessary. The commit message doesn't address this. Why do these images work without any sandbox changes?

shallawa replied:
You are right I added the needed code to enable software image decoding for AVIF in `WebProcess::platformInitializeWebProcess()`

litherum commented on Source/WebCore/platform/graphics/cg/ImageDecoderCG.cpp:
```diff
@@ -80,6 +80,10 @@ static RetainPtr<CFMutableDictionaryRef> createImageSourceOptions()
 #endif
 
+#if HAVE(IMAGE_RESTRICTED_DECODING) && HAVE(AVIF)
+    CFDictionarySetValue(options.get(), kCGImageSourceEnableRestrictedDecoding, kCFBooleanTrue);
```
I'm a bit confused by this. This line doesn't seem to be AVIF-specific, but it's behind `HAVE(AVIF)`.

Does this line have repercussions beyond just supporting AVIF?

shallawa replied:
You are right. I deleted this new code and I changed the code above it to enable the restricted (software) image decoding if HEIF or AVIF are enabled in WebKit.

litherum commented on Source/WebCore/platform/graphics/cg/ImageDecoderCG.cpp:
```diff
@@ -76,7 +76,7 @@ static RetainPtr<CFMutableDictionaryRef> createImageSourceOptions()
 #if HAVE(IMAGE_RESTRICTED_DECODING) && USE(APPLE_INTERNAL_SDK)
-    if (ImageDecoderCG::decodingHEICEnabled())
+    if (ImageDecoderCG::decodingHEICEnabled() || ImageDecoderCG::decodingAVIFEnabled())
         CFDictionarySetValue(options.get(), kCGImageSourceEnableRestrictedDecoding, kCFBooleanTrue);
```
Still not super clear if this line has any repercussions beyond HEIC or AVIF, but okay.

shallawa committed:
[Cocoa] Support AVIF images for macOS Ventura and iOS 16
https://bugs.webkit.org/show_bug.cgi?id=241904
rdar://95742091

Reviewed by Myles C. Maxfield.

Add the mime type and the UTI of the AVIF to the list of the allowed image
formats. The system frameworks will be used to render the AVIF images on macOS
Ventura and iOS 16. Because of sand-boxing limitations, software decoding has
to be used for AVIF images.

* LayoutTests/platform/mac/TestExpectations:
* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/loader/cache/CachedResourceRequest.cpp:
(WebCore::acceptHeaderValueForImageResource):
* Source/WebCore/platform/MIMETypeRegistry.cpp:
* Source/WebCore/platform/graphics/cg/ImageDecoderCG.cpp:
(WebCore::createImageSourceOptions):
* Source/WebCore/platform/graphics/cg/UTIRegistry.cpp:
(WebCore::defaultSupportedImageTypes):

Canonical link: https://commits.webkit.org/251850@main
